### PR TITLE
QOL improvemenets:  Codewriter, and installing an .editorconfig

### PR DESCRIPTION
**Pull request by Lucasharskamp**
1. I installed an .editorconfig based on the default Microsoft C# Code Conventions. It throws out suggestions (code style) or build warnings (for potential issues), and VS/Resharper automatically picks it up for any project in the solution.

2. Set up Documentation outputs for GlueFormsCore, so XML documentation is not only useful for nuget packages/DLL usage, but also auto-documentation should you ever want it.

3. Codewriter.cs:
- Added XML documentation where needed
- Made sure all similar parameters use the name naming convention and order of parameters (Codeblock, Element) so that it remains the same throughout the entire project.
- Removed unused private/internal code
- Removed very old comments with code
- Improved type-check assignments
- Improved implicit casts, costs no performance but makes the code more reliable.
- Replaced home-made string concenation solutions with String.Join for reliability and performance.
- Added OrdinalIgnoreCase for string equality checks where needed, to prevent culture-based string check mishaps and improve performance.

### CODE REVIEW

File: FRBDK/Glue/Glue/CodeGeneration/ElementComponentCodeGenerator.cs [hunk 1]
```diff
@@ -86,14 +86,14 @@ public virtual ICodeBlock GenerateDestroy(ICodeBlock codeBlock, IElement element
 
         #endregion
 
-        public virtual void GenerateActivityEditMode(ICodeBlock codeBlock, GlueElement element) {  }
+        public virtual void GenerateActivityEditMode(ICodeBlock codeBlock, IElement element) {  }
```
Comment (vchelaru) [inline FRBDK/Glue/Glue/CodeGeneration/ElementComponentCodeGenerator.cs]: This should remain GlueElement not IElement - we are slowly moving away from IElement.

File: FRBDK/Glue/Glue/CodeGeneration/ElementComponentCodeGenerator.cs [hunk 2]
```diff
@@ -86,14 +86,14 @@ public virtual ICodeBlock GenerateDestroy(ICodeBlock codeBlock, IElement element
...
 
         public virtual ICodeBlock GenerateActivity(ICodeBlock codeBlock, IElement element)
         {
             return codeBlock;
         }
 
-        public virtual void GeneratePauseIgnoringActivity(ICodeBlock codeBlock, GlueElement element) { }
+        public virtual void GeneratePauseIgnoringActivity(ICodeBlock codeBlock, IElement element) { }
```
Comment (vchelaru) [inline FRBDK/Glue/Glue/CodeGeneration/ElementComponentCodeGenerator.cs]: Same as above - continue using GlueElement.

File: FRBDK/Glue/Glue/CodeGeneration/CustomVariableCodeGenerator.cs [hunk 3]
```diff
@@ -833,10 +831,10 @@ public static ICodeBlock AppendAssignmentForCustomVariableInElement(ICodeBlock c
             //if (!customVariable.SetByDerived && 
             else if (customVariable.DefaultValue != null && // if it's null the user doesn't want to change what is set in the file or in the source object
                 !customVariable.IsShared && // no need to handle statics here because they're always defined in class scope
-                !IsVariableTunnelingToDisabledObject(customVariable, glueElement)
+                !IsVariableTunnelingToDisabledObject(customVariable, saveObject)
```
Comment (vchelaru) [inline FRBDK/Glue/Glue/CodeGeneration/CustomVariableCodeGenerator.cs]: "saveObject" is a general term for any class that has gets serialized. We should be explicit when we can, so glueElement is better in this context.

File: FRBDK/Glue/Glue/CodeGeneration/StateCodeGenerator.cs [hunk 4]
```diff
@@ -530,21 +530,19 @@ private static Dictionary<string, StateSaveCategory> GetAllStateCategoryNames(IE
...
 
 
-        private static bool IsStateDefinedInBase(IElement element, string enumName)
+        private static bool IsStateDefinedInBase(GlueElement element, string enumName)
         {
-            bool toReturn = false;
             if (string.IsNullOrEmpty(element.BaseElement))
             {
-                toReturn = false;
+                return false;
```
Comment (vchelaru) [inline FRBDK/Glue/Glue/CodeGeneration/StateCodeGenerator.cs:537]: No "early outs" unless explicitly at the very beginning of the method. toReturn is intentional.

**Comment by vchelaru** [comment]
Changes from IElement to GlueElement may also cascade into live edit mode. Communication from game <-> glue depends on the signatures of methods, so changing the method in Glue may break communication between game<->glue.

Therefore, before this can be merged, testing must be done in live edit to verify that it continues to work and a lot of the basic functionality is not broken. This includes:

* Selecting a Screen/Entity in Glue and seeing the change in game
* Selecting an instance (object) in Glue, seeing it highlight in game. Test this in screens and entities
* Selecting an instance in game and seeing it highlight in the tree view
* Copy paste in glue and seeing it update in game
* Copy paste in game, seeing it update in glue
* Move objects in game, seeing variables update in Glue

I realize that this testing is considerable, but merging the change without this testing introduces the possibility of breaking people's games and breaking live edit mode, so it must be done.

File: FRBDK/Glue/Glue/CodeGeneration/StateCodeGenerator.cs (hunk 4 above)
Comment (Lucasharskamp) [reply FRBDK/Glue/Glue/CodeGeneration/StateCodeGenerator.cs:537]: it is at the very beginning of the method. See line 533-535

**Comment by vchelaru** [comment]
Here's a doc on live edit which will be helpful:

https://docs.flatredball.com/flatredball/glue-reference/enable-live-edit

File: FRBDK/Glue/Glue/CodeGeneration/StateCodeGenerator.cs (hunk 4 above)
Comment (vchelaru) [reply FRBDK/Glue/Glue/CodeGeneration/StateCodeGenerator.cs:537]: I found a coding document, cleaned it up. See the //////////early out//////// section for examples:

https://docs.flatredball.com/flatredball/contributing/general-programming-flatredball-programming-style-guide#patterns

**Comment by vchelaru** [comment]
Unfortunately the PR has gotten too big. It includes:

* IElement -> GlueElement 
* Various variable name changes on codeblock
* A style guide file (which has not been approved or reviewed, nor asked for)

Each one of these should be its own separate PR rather than mixing them. It makes reviewing far too difficult and error prone.

**Comment by Lucasharskamp** [comment]
> Unfortunately the PR has gotten too big. It includes:
> 
> * IElement -> GlueElement
> * Various variable name changes on codeblock
> * A style guide file (which has not been approved or reviewed, nor asked for)
> 
> Each one of these should be its own separate PR rather than mixing them. It makes reviewing far too difficult and error prone.

We talked about the style guide at length in the discord and you asked for the GlueElement refactor, but agreed, we can split these up. Also, the generated code in NetStandard and this branch was 100% equal down to the last char.  Tested with FreeFileSync to test two directories (one containing .generated.*.cs files from NetStandard and the other from this branch)

I'll make a small PR for the style guide so you can see it in action on your VS.

**Comment by vchelaru** [comment]
As discussed in discord, this PR is being closed because it is too big and should be broken down. the IElement -> GlueElement refactor would be useful to do, but it must be done in steps, first doing changes not related to GlueCommands or GlueState and then later changing those.

**Pull request closed without merging**